sys/riotboot/flashwrite.c

```diff
@@ -150,6 +151,33 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
     return 0;
 }
 
+int riotboot_flashwrite_invalidate(int slot)
+{
+    if (riotboot_slot_numof == 1){
+        LOG_WARNING(LOG_PREFIX "Only one slot configured\n");
+        return -1;
+    }
+    if ((riotboot_slot_validate(riotboot_slot_other()) != 0) || (riotboot_slot_validate(riotboot_slot_current()) != 0)){
+        LOG_INFO(LOG_PREFIX "There will be no valid images to run after reboot\n");
+        return -2;
```

Comment:
This suggestion is up for discussion, the logic you have currently is requiring that both slots be valid, but IMO if you try to erase an already invalid slot, should this be allowed? I don't think this should return an error, so I adapted the logic that way.

```suggestion
    if (slot == riotboot_slot_current()) {
        if (!riotboot_slot_validate(riotboot_slot_other())) {
            LOG_WARNING(LOG_PREFIX "abort, slot %d is the only valid image\n", slot);
            return -2;
        }
    } else {
        if (!riotboot_slot_validate(riotboot_slot_current())) {
            LOG_WARNING(LOG_PREFIX "abort, slot %d is the only valid image\n", slot);
            return -2;
        }
```

Reply:
> I would not prefer to allow running the invalidation process twice on the same slot ( to invalidate the invalid).

Why? From a user perspective you would try to erase and get an error code when there is no reason for one. If you don't want to invalidate an already invalid slot than the user can implement that check IMO and avoid calling `riotboot_flashwrite_invalidate` altogether.

Reply:
> An action that I would suggest is to erase the flashpage of the invalid slot, only after reboot.

I don't why erasing the flashpage is related here.

Reply:
> Why? From a user perspective you would try to erase and get an error code when there is no reason for one. If you don't want to invalidate an already invalid slot than the user can implement that check IMO and avoid calling `riotboot_flashwrite_invalidate` altogether.
>
I just wanted to avoid alike "NOP" state, that's the reason for the selected slot validity check. But as you said, I agree with the primary validity check from the user and thus not calling the `riotboot_flashwrite_invalidate` altogether.
I would just suggest this code for that logic, just to make the code shorter IMHO:
```suggestion
    if (riotboot_slot_validate (1 - slot) != 0) {
        LOG_WARNING(LOG_PREFIX "abort, slot %d is the only valid image\n", slot)
        return -2;
```
> 
> I don't why erasing the flashpage is related here.
>
No it was just a suggestion of a potential replacement to invalidate an invalid slot, I know it is related to another module.

Reply:
But in your suggestion you only validate one slot or am I missing something? Feel free to push your suggested change I can always comment on that after and its easier to view :)

Reply:
Yes it's checking only the slot that has not been chosen to invalidate